pyverbs/dmabuf.pyx

```diff
@@ -0,0 +1,85 @@
+
+#cython: language_level=3
+
+import weakref
+
+from os import open, close, O_RDWR
+from fcntl import ioctl
+from struct import pack_into, unpack
+from pyverbs.base cimport close_weakrefs
+from pyverbs.mr cimport DmaBufMR
+
+cdef extern from "drm/drm.h":
+    cdef int DRM_IOCTL_MODE_CREATE_DUMB
+    cdef int DRM_IOCTL_MODE_MAP_DUMB
+    cdef int DRM_IOCTL_MODE_DESTROY_DUMB
+    cdef int DRM_IOCTL_PRIME_HANDLE_TO_FD
+
+cdef class DmaBuf:
+    def __init__(self, size, unit=0):
+        """
+        Allocate DmaBuf object from a GPU device. This is done through the
+        DRI device interface (/dev/dri/card*). Usually this requires the
+        effective user id being root or being a member of the 'video' group.
+        :param size: The size (in number of bytes) of the buffer.
+        :param unit: The unit number of the GPU to allocate the buffer from.
+        :return: The newly created DmaBuf object on success.
+        """
+        self.dmabuf_mrs = weakref.WeakSet()
+        self.dri_fd = open('/dev/dri/card'+str(unit), O_RDWR)
+
+        # create a dumb scanout buffer:
+        #     struct drm_mode_create_dumb {
+        #        __u32 height;
+        #        __u32 width;
+        #        __u32 bpp;
+        #        __u32 flags;
+        #        __u32 handle; /* output */
+        #        __u32 pitch; /* output */
+        #        __u64 size; /* output */
+        #    };
+        create_dumb = bytearray(32)
+        pack_into('=iiiiiiq', create_dumb, 0, 1, size, 8, 0, 0, 0, 0)
+        ioctl(self.dri_fd, DRM_IOCTL_MODE_CREATE_DUMB, create_dumb)
+        a, b, c, d, self.handle, e, self.size = unpack('=iiiiiiq', create_dumb)
+
+        # get dma-buf file descriptor from a PRIME handle:
+        #     struct drm_prime_handle {
+        #         __u32 handle;
+        #         __u32 flags;
+        #         __s32 fd; /* output */
+        #     };
+        prime_handle = bytearray(12)
+        pack_into('=iii', prime_handle, 0, self.handle, O_RDWR, 0)
+        ioctl(self.dri_fd, DRM_IOCTL_PRIME_HANDLE_TO_FD, prime_handle)
+        a, b, self.fd = unpack('=iii', prime_handle)
+
+        # set up for mmap of a dumb scanout buffer:
+        #     struct drm_mode_map_dumb {
+        #         __u32 handle;
+        #         __u32 pad;
+        #         __u64 offset; /* output */
+        #     };
+        map_dumb = bytearray(16)
+        pack_into('=iiq', map_dumb, 0, self.handle, 0, 0)
+        ioctl(self.dri_fd, DRM_IOCTL_MODE_MAP_DUMB, map_dumb);
+        a, b, self.map_offset = unpack('=iiq', map_dumb);
+
+    def __dealloc__(self):
+        close_weakrefs([self.dmabuf_mrs])
+
+        # destroy a dumb scanout buffer:
+        #     struct drm_mode_destroy_dumb {
+        #         __u32 handle;
+        #     };
+        destroy_dumb = bytearray(4)
+        pack_into('=i', destroy_dumb, 0, self.handle)
+        ioctl(self.dri_fd, DRM_IOCTL_MODE_DESTROY_DUMB, destroy_dumb)
+        close(self.dri_fd)
```

Comment:
before destroying please check if self.dri_fd is NULL. and after closing assign the self.dri_fd to be NULL (to prevent double deallocation - in case the GC done that twice or both the user and GC try to close. it shouldn't happen, but unfortunately sometimes cython has its "magic")